vircadia/vircadia
# Material improvements 2: reference textures/materials by entity ID

builds on https://github.com/vircadia/vircadia/pull/1076, diff here: https://github.com/HifiExperiments/project-athena/compare/materials...materials2

implements https://github.com/vircadia/vircadia/issues/114 (but maybe needs more discussion?)

this is either a cool feature or incredibly stupid, I can't tell which

basically, allows you to use an entity ID as the URL to a texture or material.  **textures can currently only refer to image or web entities, and materials can currently only refer to material entities (or an entity with a material entity applied)**

for example: for textures, you can make a material entity with the following JSON:
```
{
    materials: {
        "albedoMap": <ID of web or image entity>
    }
}
```
and the material entity will copy that texture.  this lets you apply web surfaces to other entities/avatars!  it also lets you have multiple web views without the cost of multiple web entities.  however, only the original web entity will accept mouse events.

and for materials, you can either just use a material entity ID as another entity's materialURL, or you can put the ID in the materials field, like:
```
{
    materials: <ID of material entity>
}
```
and you can still have multiple materials:
```
{
    materials: [
        <ID of material entity>,
        {
            name: "blah",
            albedo: [1, 0, 0]
        }
    ]
}
```
(in which case, you can still refer to each of them in the materialURL by name, and for entity IDs, the name is just the ID)

the only downside is that currently, the entity that you're referencing must also be loaded.  this means you can't make a material to another material that is halfway across the world, it needs to be pretty close so that the entity server sends it etc.  I am still thinking about how to fix this

also fixes selection by name not working when the URL is materialData

Test plan:
- Create a web entity, an image entity, and a material entity with materialURL = materialData
- Set the materialData to:
```
{
    materials: {
        "albedoMap": <ID of the web entity>
    }
}
```
- Confirm that as the web surface changes, the material updates to match it.  Apply the material to a cube or model and confirm that the web surface is applied to that entity.
- In the material entity, change the ID to the image entity.  Confirm that the material entity now uses that texture.  Change the imageURL to a different image and confirm that it updates the material.
- Confirm that procedurals can accept entity IDs as input channels.  For example, you can make a material entity with materialData:
```
{
  "materials": {
    "model": "hifi_shader_simple",
    "procedural": {
      "version": 2,
      "fragmentShaderURL": "https://gist.github.com/HifiExperiments/51edf2e107cbeab18474416b751c1e07/raw/1817d7a2f5bf73978b1a759ecac93e936fc75d1d/ProceduralTexture.fs",
      "channels": [<id of web entity>]
    }
  }
}
```
- Create a material entity with materialURL "https://hifi-content.s3.amazonaws.com/samuel/TestMaterial.json".  This should be a normal material with a texture applied.  Call this materialA;
- Create a material entity with materialData:
```
{
  "materialVersion": 1,
  "materials": [
    {
      "name": "red",
      "model": "hifi_pbr",
      "albedo": [1, 0, 0]
    },
    {
      "name": "green",
      "model": "hifi_pbr",
      "albedo": [0, 1, 0]
    },
    {
      "name": "blue",
      "model": "hifi_pbr",
      "albedo": [0, 0, 1]
    },
    "<the ID of materialA>"
  ]
}
```
- Confirm that you can select either red, blue, or green by appending "?red", "?blue", or "?green" to the materialURL (this is broken sometimes in master right now)
- Confirm that you can select the ID by name by changing the materialURL to `materialData?<the ID of materialA>`
- Confirm that you can control the texture transforms of materialA and the last material separately.
